restlicsharpclient/restliclient/util/DataMapDeserializationConverter.cs

```diff
@@ -0,0 +1,37 @@
+﻿using System;
+using System.Collections.Generic;
+using Newtonsoft.Json.Converters;
+using Newtonsoft.Json;
+
+namespace restlicsharpclient.restliclient.util
+{
+    class DataMapDeserializationConverter : CustomCreationConverter<object>
+    {
+        public override object Create(Type objectType)
+        {
+            if (objectType == typeof(Array))
+                return new List<object>();
+            else
+                return new Dictionary<string, object>();
+        }
+
+        public override bool CanConvert(Type objectType)
+        {
+            // in addition to handling Dictionary<string, object>
+            // we want to handle the deserialization of dict value
+            // which is of type object
+            return objectType == typeof(object) || base.CanConvert(objectType);
+        }
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            if (reader.TokenType == JsonToken.StartArray)
+                return base.ReadJson(reader, typeof(Array), existingValue, serializer);
+            else if (reader.TokenType == JsonToken.StartObject || reader.TokenType == JsonToken.Null)
```

Comment:
Just curious, for JsonToken.StartObject, you don't need to pass in `typepf(Dictionary)` here as you did for Array? I thought that after deserialization, we get a Dictionary<string, object> where nested record will become another Dictionary inside the Dictionary.